### PR TITLE
Fix servers?dsID=<> route to return mids correctly

### LINKED ISSUE #5493: GET /servers?dsId={{dsId}} not filtering out cache servers (edges/mids) w/o required server capabilities
## I'm submitting a ...
-  bug report

## Traffic Control components affected ...
-  Traffic Ops

## Current behavior:
GET /servers?dsId={{dsId}} is returning mids that do not have the ds's required server capabilities.

## Expected behavior:
If a delivery service requires one or more server capabilities, the results of GET /servers?dsId={{dsId}} should never include cache servers (edges or mids) that do not have all of the ds's required server capabilities. ORG servers are different thanks to MSO and are returned regardless.

## Minimal reproduction of the problem with instructions:
- create a ds that employs both the edge and mid tier. i.e. a DNS delivery service
- add a required server capability to the ds
- add the same server capability to a handful of edge servers (ensure those edge servers are in a cache group that is parented by a mid cache group) and assign those edges to the ds
- GET /servers?dsId={{dsId}} and notice that mids are returned when mids were never assigned the server capability

## Anything else:
Note: this does not seem to affect the behavior of "server capabilities" which was designed to include/exclude cache servers (edges/mids) from the crconfig (snapshot) and relevant ats config files (parent.config).

### PR DESCRIPTION
**Pull request by srijeet0406**
## What does this PR (Pull Request) do?

- [x] This PR fixes #5493  

## Which Traffic Control components are affected by this PR?

- Traffic Ops
- CI tests

## What is the best way to verify this PR?

create a ds that employs both the edge and mid tier. i.e. a DNS delivery service
add a required server capability to the ds
add the same server capability to a handful of edge servers (ensure those edge servers are in a cache group that is parented by a mid cache group) and assign those edges to the ds
GET /servers?dsId={{dsId}} and make sure that the mids are not returned because they dont have the server capability

## If this is a bug fix, what versions of Traffic Control are affected?

- master

## The following criteria are ALL met by this PR

- [x] This PR includes tests 
- [x] This PR does not include documentation 
- [x] This PR does not include an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

## Additional Information

NOTE: This si the part that is necessary to fix this behavior in the 5.1 milestone. I'll create another PR for API 4.0(TC 6.0).

### CODE REVIEW

File: traffic_ops/traffic_ops_golang/server/servers.go [hunk 1]
```diff
@@ -1237,6 +1248,22 @@ func getMidServers(edgeIDs []int, servers map[int]tc.ServerV40, dsID int, cdnID
 		FROM deliveryservice d
 		WHERE d.id = :ds_id) IS NULL
 	`
+		query = selectQuery + `
+	WHERE (SELECT ARRAY_AGG(ssc.server_capability) FROM server_server_capability ssc WHERE ssc."server" = ANY(
+	` + q + `)) @> (SELECT ARRAY_AGG(drc.required_capability) FROM deliveryservices_required_capability drc WHERE drc.deliveryservice_id = :ds_id)`
```
Comment (zrhoffman) [inline traffic_ops/traffic_ops_golang/server/servers.go:1253]: Can this query be formatted nicer so we can tell what's going on here? Commenting the query would not hurt, either.

**Review by zrhoffman — CHANGES_REQUESTED** [review]
When the Delivery Service has Required Capabilities, I am getting servers returned for Edge Cache Groups that do not have the DS's Required Capability.

Steps to reproduce:
1. Start CDN in a Box
2. Unassign the *demo1-top* Topology from the *demo1* Delivery Service
3. Assign the *edge*  Server to the *demo1* Delivery Service

Response for GET https://trafficportal.infra.ciab.test/api/4.0/servers?dsId=1 is as expected:
```json
{
  "response": [
    {
      "hostName": "edge",
      "cachegroup": "CDN_in_a_Box_Edge"
    },
    {
      "hostName": "origin",
      "cachegroup": "CDN_in_a_Box_Origin"
    }
  ],
  "summary": {
    "count": 2
  }
}
```

4. Assign Cache Group *CDN_in_a_Box_Mid-01* as a Parent Cache Group of Cache Group *CDN_in_a_Box_Edge*.

Response for GET https://trafficportal.infra.ciab.test/api/4.0/servers?dsId=1 contains servers for Cache Group *CDN_in_a_Box_Edge* that were not previously included:
```json
{
  "response": [
    {
      "hostName": "edge",
      "cachegroup": "CDN_in_a_Box_Edge"
    },
    {
      "hostName": "origin",
      "cachegroup": "CDN_in_a_Box_Origin"
    },
    {
      "hostName": "mid-01",
      "cachegroup": "CDN_in_a_Box_Mid-01"
    },
    {
      "hostName": "mid-02",
      "cachegroup": "CDN_in_a_Box_Mid-02"
    },
    {
      "hostName": "trafficmonitor",
      "cachegroup": "CDN_in_a_Box_Edge"
    },
    {
      "hostName": "trafficrouter",
      "cachegroup": "CDN_in_a_Box_Edge"
    }
  ],
  "summary": {
    "count": 6
  }
}
```

**Comment by srijeet0406** [comment]
I had forgotten to add a `WHERE` clause, should be fixed now.

**Review by zrhoffman — CHANGES_REQUESTED** [review]
I'm noticing that a Server without all of the Delivery Service's Required Capabilities is included in the response.

Steps to reproduce:
1. Start CDN in a Box
2. Unassign the *demo1-top* Topology from the *demo1* Delivery Service
3. Assign Server *edge* to the *demo1* Delivery Service
4. Assign Cache Group *CDN_in_a_Box_Mid-01* as a Parent Cache Group of Cache Group *CDN_in_a_Box_Edge*.
5. Move Server *mid-02* to Cache Group *CDN_in_a_Box_Mid-01*
6. Remove the *HDD* Server Capability from Server *mid-02* (*HDD* is a Required Capability of the *demo1* Delivery Service)

```jsonc
// GET https://trafficportal.infra.ciab.test/api/4.0/servers?dsId=1
{
  "response": [
    {
      "hostName": "edge",
      "cachegroup": "CDN_in_a_Box_Edge"
    },
    {
      "hostName": "origin",
      "cachegroup": "CDN_in_a_Box_Origin"
    },
    {
      "hostName": "mid-02",
      "cachegroup": "CDN_in_a_Box_Mid-01"
    },
    {
      "hostName": "mid-01",
      "cachegroup": "CDN_in_a_Box_Mid-01"
    }
  ],
  "summary": {
    "count": 4
  }
}
```

**Comment by srijeet0406** [comment]
Should be fixed now

**Review by zrhoffman — APPROVED** [review]
This all works!

**Pull request merged**